src/array.c

```diff
@@ -629,156 +642,408 @@ static size_t limit_overallocation(jl_array_t *a, size_t alen, size_t newlen, si
+    if (sz <= n)
+        return;
+    size_t inc = sz - n;
+    jl_array_grow_end(a, inc);
+    a->nrows = n;
 #ifdef STORE_ARRAY_LEN
-    a->length -= dec;
+    a->length = n;
 #endif
-    a->nrows -= dec;
+}
 
-    // make sure offset doesn't grow forever due to deleting at beginning
-    // and growing at end
-    size_t newoffs = offset;
-    if (offset >= 13*a->maxsize/20) {
-        newoffs = 17*(a->maxsize - a->nrows)/100;
+JL_DLLEXPORT jl_array_t *jl_array_copy(jl_array_t *ary)
+{
+    size_t elsz = ary->elsize;
+    jl_array_t *new_ary = _new_array_(jl_typeof(ary), jl_array_ndims(ary),
+                                      &ary->nrows, !ary->flags.ptrarray, elsz);
+    memcpy(new_ary->data, ary->data, jl_array_len(ary) * elsz);
+    return new_ary;
+}
+
+// Copy element by element until we hit a young object, at which point
+// we can
+static NOINLINE ssize_t jl_array_ptr_copy_forward(jl_value_t *owner,
+                                                  void **src_p, void **dest_p,
+                                                  ssize_t n)
+{
+    for (ssize_t i = 0; i < n; i++) {
+        void *val = src_p[i];
+        dest_p[i] = val;
+        // `val` is young or old-unmarked
+        if (val && !(jl_astaggedvalue(val)->bits.gc & GC_MARKED)) {
+            jl_gc_queue_root(owner);
+            return i;
+        }
     }
-#ifdef _P64
-    while (newoffs > (size_t)((uint32_t)-1)) {
-        newoffs = newoffs/2;
+    return n;
+}
+
+static NOINLINE ssize_t jl_array_ptr_copy_backward(jl_value_t *owner,
+                                                   void **src_p, void **dest_p,
+                                                   ssize_t n)
+{
+    for (ssize_t i = 0; i < n; i++) {
+        void *val = src_p[n - i - 1];
+        dest_p[n - i - 1] = val;
+        // `val` is young or old-unmarked
+        if (val && !(jl_astaggedvalue(val)->bits.gc & GC_MARKED)) {
+            jl_gc_queue_root(owner);
+            return i;
+        }
     }
-#endif
-    if (newoffs != offset) {
-        size_t anb = a->nrows*es;
-        size_t delta = (offset - newoffs)*es;
-        a->data = (char*)a->data - delta;
-        memmove(a->data, (char*)a->data + delta, anb);
+    return n;
+}
+
+// Unsafe, assume inbounds and that dest and src have the same eltype
+// `doffs` and `soffs` are zero based.
+JL_DLLEXPORT void jl_array_ptr_copy(jl_array_t *dest, void **dest_p,
+                                    jl_array_t *src, void **src_p, ssize_t n)
+{
+    assert(dest->flags.ptrarray && src->flags.ptrarray);
+    jl_value_t *owner = jl_array_owner(dest);
+    // Destination is old and doesn't refer any young object
+    if (__unlikely(jl_astaggedvalue(owner)->bits.gc == GC_OLD_MARKED)) {
+        jl_value_t *src_owner = jl_array_owner(src);
+        // Source is young or might refer young objects
+        if (!(jl_astaggedvalue(src_owner)->bits.gc & GC_OLD)) {
+            ssize_t done;
+            if (dest_p < src_p) {
```

Comment:
Sure (without the `sizeof(jl_value_t*)`).... I would expect the reverse copying to be easily recognized by the prefetcher though....